### PR TITLE
chore: finish proposal job

### LINKED ISSUE #1330: Find a way to reduce finishProposal job complexity

### PR DESCRIPTION
**Pull request by 1emu**
Closes #1330

### CODE REVIEW

File: src/back/routes/proposal.ts [hunk 1]
```diff
@@ -564,7 +565,10 @@ export async function updateProposalStatus(req: WithAuth<Request<{ proposal: str
     NotificationService.grantProposalEnacted(proposal)
   }
 
-  ProposalService.commentProposalUpdateInDiscourse(id)
+  const updatedProposal: ProposalAttributes | undefined = await ProposalModel.findOne<ProposalAttributes>({
```
Comment (andyesp) [inline src/back/routes/proposal.ts:568]: Isn't the type being inferred from the `findOne<ProposalAttributes>`? Maybe `findOne<ProposalAttributes | undefined>`?

File: src/entities/Proposal/jobs.ts [hunk 2]
```diff
@@ -22,85 +32,42 @@ export async function activateProposals(context: JobContext) {
...
-async function getProposalsWithOutcome(proposals: ProposalAttributes[], context: JobContext) {
-  const pendingProposalsWithOutcome = []
-
+async function getProposalsVotingResult(proposals: ProposalAttributes[]) {
+  const proposalsWithVotingResult: ProposalVotingResult[] = []
   for (const proposal of proposals) {
-    // TODO: New proposal status could be calculated here and added to outcome.
-    // E.g: OutcomeObject: { ...outcome, newProposalStatus: OutOfBudget }
-    const outcome = await calculateOutcome(proposal, context)
+    const outcome = await calculateVotingResult(proposal)
```
Comment (andyesp) [inline src/entities/Proposal/jobs.ts:61]: Rename this to... `votingResult`?

File: src/entities/Proposal/jobs.ts [hunk 3]
```diff
@@ -236,3 +180,49 @@ export async function publishBids(context: JobContext) {
...
 }
+
+async function updateProposalsAndBudgets(proposalsWithOutcome: ProposalWithOutcome[], budgetsWithUpdates: Budget[]) {
+  const pool = new Pool({ connectionString: process.env.CONNECTION_STRING })
+  const client = await pool.connect()
+
+  try {
+    await client.query('BEGIN')
+
+    if (proposalsWithOutcome.length === 0) return
```
Comment (andyesp) [inline src/entities/Proposal/jobs.ts:191]: We should return before opening a new pool connection, at the start of the function.

File: src/services/BadgesService.ts [hunk 4]
```diff
@@ -269,4 +270,19 @@ export class BadgesService {
...
+  static giveFinishProposalBadges(proposalsWithOutcome: ProposalWithOutcome[]) {
+    const passedProposals = proposalsWithOutcome.filter((proposal) => proposal.newStatus === ProposalStatus.Passed)
+    inBackground(async () => {
+      try {
+        await BadgesService.giveLegislatorBadges(passedProposals)
+      } catch (error) {
+        ErrorService.report('Error while attempting to give badges', {
+          error,
+          category: ErrorCategory.Badges,
+          passedGovernanceProposals: passedProposals,
```
Comment (andyesp) [inline src/services/BadgesService.ts:283]: `passedProposals` includes all proposal types here as the filtering is done in `giveLegislatorBadges`. You could rename `passedGovernanceProposals` to `passedProposals` or you could move the governance type filter to this function and remove it from the `giveLegislatorBadges`.

File: src/services/DiscordService.ts [hunk 5]
```diff
@@ -237,10 +238,10 @@ export class DiscordService {
     }
   }
 
-  static finishProposal(id: string, title: string, outcome: string, winnerChoice?: string) {
+  static finishProposal(id: string, title: string, outcome: ProposalStatus, winnerChoice?: string) {
     if (DISCORD_SERVICE_ENABLED) {
       inBackground(async () => {
-        const action = `Proposal has ended with outcome ${outcome}`
+        const action = `Proposal has ended with outcome ${outcome.toUpperCase().replaceAll('_', ' ')}`
```
Comment (andyesp) [inline src/services/DiscordService.ts:244]: We are also doing this replace in `CategoryPill`. Maybe we could extract all this to a function and reuse it in both files. (with toUpperCase included)

File: src/services/DiscourseService.ts [hunk 6]
```diff
@@ -148,4 +153,28 @@ export class DiscourseService {
       return null
     }
   }
+
+  static commentUpdatedProposal(updatedProposal: ProposalAttributes) {
+    inBackground(async () => {
+      // TODO: votes are not necessary in all cases, they could be fetched only when needed
```
Comment (andyesp) [inline src/services/DiscourseService.ts:159]: When is it not necessary?

Comment (andyesp) [reply src/services/DiscourseService.ts:159]: Oh ok I see the check is in getUpdateMessage. Maybe I'd move the `// TODO:` to that file (It is asking for a refactor)

**Review by ncomerci — APPROVED** [review]
Lovely 💕

**Review by andyesp — APPROVED** [review]

File: src/entities/Proposal/jobs.ts [hunk 7]
```diff
@@ -236,3 +202,48 @@ export async function publishBids(context: JobContext) {
...
+          proposalStatus
+        )
+
+        if (query !== null) {
+          proposalUpdateQueriesByStatus.push(query)
+        }
+      }
+    })
+
+    const budgetUpdateQueries = BudgetService.getBudgetUpdateQueries(budgetsWithUpdates)
```
Comment (andyesp) [inline src/entities/Proposal/jobs.ts:229]: Shouldn't we skip this if `budgetsWithUpdates.length === 0`?

File: src/entities/Proposal/jobs.ts [hunk 8]
```diff
@@ -236,3 +202,48 @@ export async function publishBids(context: JobContext) {
...
+        const query = ProposalModel.getFinishProposalQuery(
+          proposalsToUpdate.map(({ id }) => id),
+          proposalStatus
+        )
+
+        if (query !== null) {
+          proposalUpdateQueriesByStatus.push(query)
+        }
+      }
+    })
```
Comment (andyesp) [inline src/entities/Proposal/jobs.ts:227]: I'd move this to a `getFinishProposalQueries` in the ProposalService.

File: src/entities/Proposal/jobs.ts [hunk 9]
```diff
@@ -121,51 +89,66 @@ function hasCustomOutcome(type: ProposalType) {
...
+        ErrorService.report('Could not find voting results for pending proposal', {
+          proposal: pendingProposal,
+          category: ErrorCategory.Job,
+        })
+      }
+    })
+  }
+}
+
+async function categorizeProposals(
```
[the remark this answers is not in the document]
Comment (1emu) [reply src/entities/Proposal/jobs.ts:112]: `processProposalsAndBudgets`? `prepareProposalsAndBudgetsUpdates`?

File: src/entities/Proposal/jobs.ts (hunk 7 above)
Comment (1emu) [reply src/entities/Proposal/jobs.ts:229]: We could check if `budgetsWithUpdates` has any difference with `currentBudgets`

**Pull request merged**